### PR TITLE
viewer: don't repeatedly show overlapping annotations

### PR DESCRIPTION
**Pull request by wjt**
See GH issue #520.

This is problematic because it will break mouseover events on the
highlight's parent, but it does make the page not freeze when you mouse
in and out of a chunk of text with hundreds of annotations.  See
<http://willthompson.co.uk/misc/annotator-freeze-on-mouseout/> for an
example without this patch.

One alternative would be to use a single span.annotation-hl with
multiple annotation-ids where currently they would be nested. ie,
transform this:

```html
<span class="annotation-hl" data-annotation-id="123">
  <span class="annotator-hl" data-annotation-id="456">
    ...
      Foo
    ...
  </span>
</span>
```

into this:

```html
<span class="annotation-hl" data-annotation-id="123 456 ...">
  Foo
</span>
```

Another alternative would be to stick with the current structure but
make the .done() callback inside _onHighlightMouseover smarter about not
repeating itself.

### CODE REVIEW

**Comment by tilgovi** [comment]
Good find. I wonder how much effort it would be to change the highlighter to coalesce spans.

**Comment by wjt** [comment]
I've actually ended up looking at the span-creation stuff anyway (for performance when annotating a range of every cell in a 5000-row table – perhaps pathological but that's where I find myself) so maybe I'll take a look…

**Comment by tilgovi** [comment]
:cake: thanks

**Comment by BigBlueHat** [comment]
@wjt is this still needed given your work on #523? Not sure how close to a fix you got with that later PR.

Thanks!

**Comment by wjt** [comment]
@BigBlueHat this is orthogonal to #523 unfortunately.

I've an alternative patch which doesn't break other mouseover handlers in the document, plus some other fixes I noticed along the way – just a mo…

**Comment by tilgovi** [comment]
@wjt bump

**Comment by wjt** [comment]
Sorry, should have been explicit that I had pushed the “alternative patch”!

Calling `$(this).parent()` repeatedly is a bit sad I guess. The other approach I could see is comparing `this` to `event.target`, but that would fail if there is other markup inside the innermost `annotator-hl`.

**Comment by tilgovi** [comment]
There shouldn't be any interior markup unless it's added outside annotator
after the highlights have been added because the highlighter wraps the text
nodes.

**Comment by wjt** [comment]
Fair enough – I guess that if something is meddling with the contents of the highlighter's elements after they are created, there may be bigger problems. Here's a version which checks `event.target`.

File: src/ui/viewer.js [hunk 1]
```diff
@@ -117,7 +117,11 @@ var Viewer = exports.Viewer = Widget.extend({
 
             $(this.options.autoViewHighlights)
                 .on("mouseover." + NS, '.annotator-hl', function (event) {
-                    self._onHighlightMouseover(event);
+                    // If there are many overlapping highlights, still only
```
Comment (tilgovi) [inline src/ui/viewer.js]: I don't think this part is necessary since the event is delegated.

Comment (tilgovi) [inline src/ui/viewer.js]: Never mind. Brain failure.

**Comment by tilgovi** [comment]
Thanks! Sorry that took me so long.

**Comment by wjt** [comment]
No prob, thanks for merging!

**Pull request merged**